pytorch_lightning/loops/batch/flexible_optimization_flow.py

```diff
@@ -0,0 +1,161 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+import logging
+from collections import OrderedDict
+from copy import copy
+from typing import Iterator, List, Optional, Tuple
+
+import torch
+
+import pytorch_lightning as pl
+from pytorch_lightning.trainer.progress import OptimizationProgress
+from pytorch_lightning.trainer.supporters import TensorRunningAccum
+from pytorch_lightning.utilities import AttributeDict
+from pytorch_lightning.utilities.batch_loop_common import (
+    check_finite,
+    check_training_step_output,
+    process_training_step_output,
+)
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.model_helpers import is_overridden
+
+log = logging.getLogger(__name__)
+
+
+class FlexibleOptimizationFlow:
```

Comment:
This should be a loop.

However, I don't believe there is a need to introduce a new loop at this stage as this new functionality is pretty straight-forward and this will increase maintenance cost in the future.

I think we should improve the `BatchLoop` API by refactoring its API and provide more user friendly utilities.

Reply:
I was heavily conflicted with this one. On one hand, the class assumes the role of the innermost loop in this setup (like `TrainingBatchLoop`). On the other hand, functionality-wise it doesn't act like a loop.

> However, I don't believe there is a need to introduce a new loop at this stage as this new functionality is pretty straight-forward and this will increase maintenance cost in the future.

Totally agree. I feel it's worthwhile introducing an internal abstraction for optimization flows (for encapsulating the logic that actually calls `training_step`) and always attached it to the end of the loop structure (currently the inner most Loop invokes `training_step`). IMHO this would improve the maintainability of different optimization flows because (1) we envision supporting more of them (2) the complexity of different flows vary greatly (e.g. the flow for automatic optimization is much more complicated than that of manual optimization) (see discussion in https://github.com/PyTorchLightning/pytorch-lightning/issues/8732#issuecomment-895081260).

cc: @awaelchli - this is an example where the component invoking `training_step` doesn't need to be a `Loop`.

Reply:
It could be in a directory `pytorch_lightning/loops/processors/flexible_optimization_flow.py`

Also, the name is very ambiguous, maybe rename it to `IteratorBatchProcessor`?

Reply:
Hey @ananthsub @yifuwang @williamFalcon,

I don't fully agree with the previous statement this can't be a loop.

The current implementation doesn't iterate over the `active optimizers` nor `split batches`, but when we had support for them, it would become a loop.

Maybe a better abstraction would be to introduce an `Optimizer Loop` and then an `IteratorBatchProcessor`.

Finally, I am slightly concerned about the direction that this PR is taking.
 
To me, it seems to be a speed optimisation feature very tailored to FB and the existence of code relying on dataloader_iter being available.
I don't believe this would benefit the entire community while it is adding an extra maintenance cost within Lightning and one could argue this feature don't meet the criteria to be integrated within Lightning core.

However, we could increase the feature scope as started within (#8700) which could benefit a larger portion of the Lightning community while providing support for researcher to access the `dataloader_iter` for `optimization purposes`.

Reply:
I strongly suggest to think about a broader design and enable this feature by the same occassion.

```py
class LightningFetcher:

    # this is own by the DataConnector

    def __init__(
        self,
        prefetch_batches: int = 1,
        dataloader,
        batch_to_device: Callable,
        profiler: "pl.profiler.base.BaseProfiler",
    ) -> None:
        self.iterator = profiled_iterator(dataloader, profiler)
        self.profiler = profiler
        self.batch_to_device = batch_to_device
        self.batches: List = []
        self.events: List = []
        self.counter: int = 0
        self.num_prefetch_batches = num_prefetch_batches

    def setup(self, dataloader, batch_to_device, profiler):
        self.dataloader = dataloader
        self.batch_to_device = batch_to_device
        self.profiler = profiler

    def __iter__(self) -> Generator:
        self.counter = 1
        if self.num_prefetch_batches == 1:
            return self.prefetch_single_batch_iterator()
        return self.prefetch_iterator()

    def add_event(self, event) -> None:
        self.events.append(event)

    def add_batch(self, batch) -> None:
        self.batches.append(batch)

    @staticmethod
    def start_record(event) -> None:
        event.record()

    def fetch_batch(self):
        return self.batches.pop(0)

    def wait(self) -> None:
        event = self.events.pop(0)
        event.wait()

    def store_on_iterator(self, iterator, iterator_state):
        ...

    def prefetch_iterator(self) -> Any:
        done = False
        while not done:
            for _ in range(self.num_prefetch_batches + 1):
                if not done:
                    try:
                        batch = next(self.iterator)
                        self.add_batch(batch)
                    except StopIteration:
                        done = True

            batch = self.fetch_batch()
            # yield last and has next
            yield self.counter, batch, done
            self.counter += 1

class InterBatchParallelism(LightningFetcher):

    def prefetch_iterator(self) -> Any:
            cuda_stream = torch.cuda.Stream()
        
            done = False
            while not done:
        
                for _ in range(self.num_prefetch_batches + 1):
                    if not done:
                        with torch.cuda.stream(cuda_stream):
                            batch_event = torch.cuda.Event()
                            self.add_event(batch_event)
                            try:
                                batch = next(self.iterator)
                                with self.profiler.profile("training_batch_to_device"):
                                    batch = self.batch_to_device(batch)
                                self.add_batch(batch)
                                self.start_record(batch_event)
                            except StopIteration:
                                done = True
        
                self.wait()
                batch = self.fetch_batch()
                # yield last and has next
                yield self.counter, batch, done
                self.counter += 1

Trainer(prefetch_batches=[0, 1, ... "auto", LightningFetcher(prefetch_batches=1))
```

Reply:
```from torch.utils.data import DataLoader
from contextlib import suppress

class DataLoaderTracker:

    def __init__(self, dataloader, prefetcher):
        self.dataloader = dataloader
        self.prefetcher = prefetcher
        self._done = False

    @property
    def done(self):
        return self._done

    def __iter__(self):
        for batch in self.dataloader:
            self.prefetcher.counter += 1
            yield batch
        self._done = True


class TrainingStepIteratorFetcher:

    def __init__(
        self,
        dataloader,
    ) -> None:
        self.dataloader = dataloader
        self.counter = 0

    def __iter__(self):
        self.tracker = DataLoaderTracker(self.dataloader, self)
        iterator = iter(self.tracker)
        while not self.tracker.done:
            yield self.counter, iterator
        raise StopIteration


dataloader = DataLoader(range(100))
prefetcher = TrainingStepIteratorFetcher(dataloader)
iterator = iter(prefetcher)
counter, dataloader_iter = next(iterator)
batch = next(dataloader_iter)
batch = next(dataloader_iter)
assert prefetcher.counter == 2
counter, dataloader_iter = next(iterator)
assert counter == 2
batch = next(dataloader_iter)
batch = next(dataloader_iter)
assert prefetcher.counter == 4
for batch in dataloader_iter:
    pass
assert prefetcher.counter == 100
assert prefetcher.tracker.done
with suppress(RuntimeError):
    counter, dataloader_iter = next(iterator)

Reply:
Will continue the discussion in https://github.com/PyTorchLightning/pytorch-lightning/issues/8867